### ISSUE TITLE
[Feature request] Custom json passed as a webhook alert

### ISSUE THREAD

**Issue by Civil** [grafana/grafana#6956]
Right now it seems not possible to send proper json (based on alert, with templating) to a webhook based alert.

We currently have our custom API to send actual alerting message, but it needs a lot of metadata in json.

**Comment by torkelo** [comment]
what do you mean proper json?

**Comment by Civil** [comment]
Sorry, not a "proper json", but a "custom" json. For example we already have an endpoint where we can POST some data and it'll do the notification and tracking. But it's different from what grafana sends and will not be very easy to adapt.

**Comment by varunsrinath** [comment]
@torkelo as per the below example JSON mentioned in http://docs.grafana.org/alerting/notifications/#webhook , how is grafana picking values for message and ruleName?

**Comment by Artanicus** [comment]
Would love to use Pushover as my notification system but not being able to define the needed json fields makes this impossible. Alternatively I could write a new notification plugin but notification plugins don't seem to be a thing yet.

**Comment by generelspaz** [comment]
I tend to agree that there are several missing pieces on the notification functionality. 

1. Need to be able to include a secret hash to enable API auth tokens
2. The ability to build the body for the json post being sent would be invaluable at providing options to use multiple platforms to send different forms of notifications and different tools to send them(text/twillio or a different service) 
3. We need the ability to set a token for the json body being sent from the alert.
Ex. Alert Name with value, Alert endpoint, alert system

Ultimately right now with the notifications you have to build a wrapper service to transform your web hooks to have any value.

**Comment by hExPY** [comment]
+1

**Comment by shiv4nsh** [comment]
:+1:

**Comment by serragnoli** [comment]
Any news on when/if the feature will be available in a near future?

**Comment by daniellee** [comment]
@serragnoli It's not planned for the next release. Roadmap [here](https://github.com/grafana/grafana/blob/master/ROADMAP.md).

**Comment by jaimegago** [comment]
@daniellee Any news on this ?

**Comment by qhh0205** [comment]
@torkelo
Does JSON data can be customized now?

**Comment by jkumar-roambee** [comment]
Required the same feature(Custom body for webhook alert notification) to integrate Microsoft team webhook integration.  +1

**Comment by D031121** [comment]
Getting this feature of custom json body would be really helpful

**Comment by dburianov** [comment]
+1

**Comment by superliyang** [comment]
+1

**Comment by maidirepelle** [comment]
+1

**Comment by whitetigle** [comment]
+1

**Comment by abbottbd** [comment]
We would also like to see this - would like to integrate Grafana with our TFS & Rundeck servers but we can't customize the webhook json so we can't trigger jobs in the other services based on alerts.

**Comment by anton-karlsson0** [comment]
+1

**Comment by rupeshreddytumu** [comment]
+1

**Comment by tmvtmv** [comment]
+1

**Comment by heytrav** [comment]
:+1:

**Comment by mikkop71** [comment]
Is this any closer to getting selected in the development phase or is this still in the backlog? I would certainly appreciate seeing this feature.

**Comment by deathwarrior60** [comment]
+1

**Comment by mtetreault** [comment]
What is the status on this? We would really need this feature (Custom body for webhook alert notification).

Having a custom body for webhook alert notification would allow us to use a prometheus alert notification webhook without the need of prometheus' alertmanager.

Is this feature, on the backlog to the complexity? I am trying to get my head around this.

**Comment by dxs34805** [comment]
We are also looking for a way to use custom json bodies in our alert messages so that we can send alert into our ServiceNow Incident Management system. Being able to customize the body would allow us to pass the necessary variables about the alert's context.

**Comment by titaneric** [comment]
Look forward to seeing this feature

**Comment by allareddysrikanth** [comment]
Badly need this feature

**Comment by titaneric** [comment]
without this feature, user HAVE to build another proxy-server for constructing user-defined JSON payload which is troublesome

**Comment by iimnurd** [comment]
need webhook payload data like : 
- timestamp, when notification triggers
- query triggers
or another custom payload data.

hopefully this feature will provide soon. thankyou  👍

**Comment by mooneym17** [comment]
Custom json or exex external script with passed in variables would be helpful

**Comment by DanielGurtne** [comment]
It would be great if you could use a webhook to trigger a job template via API on an ansible tower (awx). Or have a separate dedicated interface for it.

**Comment by ever4Kenny** [comment]
+1 for custom json

**Comment by wr-nathan** [comment]
+1 for custom json payload

**Comment by Seb-a320** [comment]
+1

**Comment by Kot-o-pes** [comment]
+1

**Comment by daynthishoho** [comment]
+1

**Comment by cody770** [comment]
+1

**Comment by Erhao** [comment]
Very much looking forward to custom JSON

**Comment by Reinhard-F** [comment]
This option would help me a lot!
+1

**Comment by BendeguzKoszticsak** [comment]
+1

**Comment by grobinson-grafana** [comment]
This feature has been discussed in the Alerting team but it is not a simple feature to build. We want to allow users to template JSON, but writing templates that produce valid JSON is difficult, and becomes more difficult when the template includes if statements and for loops. For example, in JSON the last field must not have a comma at the end.

We have not forgotten about this use case, but we are still thinking about how to deliver this feature as best as possible.

**Comment by bruno-kakele** [comment]
+1

**Comment by xiaofud** [comment]
is this feature available now?

**Comment by domi27** [comment]
This would be a very big improvement in an industrial or business environment, where you often have to trigger / notify other subsystems. 
+1

**Comment by zoic21** [comment]
+1

**Comment by lewis262626** [comment]
+1

**Comment by wuast94** [comment]
+1

**Comment by andeke07** [comment]
+1

**Comment by kyrianae** [comment]
Did someone thought about a way to do this feature ?

some times ago I used a panel button (don't remember which one, cloudspot maybe) that allow to write the payload with variables. that could fit the need on noitification.

another way:
you could add each keys in the configuration with the associated value in go tempalte style by default,
it would permits to people to suppress all fields and add all custom they want
As fixed structure in code is too restrictive

go template style allo to manage list and so one, then should fullfil a large part of need

it fix json output only but it will be great as it open the door to any json input API

**Comment by timo1707** [comment]
+1

**Comment by m-rcl** [comment]
+1

**Comment by lg-lima1** [comment]
+1

**Comment by SeaweedbrainCY** [comment]
+1

**Comment by grobinson-grafana** [comment]
Thanks for all the +1s. Just to add some background here, Grafana embeds the Prometheus Alertmanager. It offers most of the same integrations as Alertmanager, but with some additional features like support for images. There are a number of open requests to support this in Alertmanager too (https://github.com/prometheus/alertmanager/issues/1496, https://github.com/prometheus/alertmanager/pull/2184) which should give more information about the challenges of supporting custom JSON in webhook notifications.

**Comment by NicoCaldo** [comment]
+1

**Comment by joerg** [comment]
Supporting JSON bodys (`Content-type: application/json` Header) would allow using Slack Webhooks with modern and interactive Block Kit (https://api.slack.com/block-kit) messages. This is a feature we are really missing in our paid Grafana Cloud instance.

**Comment by erfantkerfan** [comment]
+1

**Comment by bestK** [comment]
Hi everyone, I've created a [grafana-webhook-converter](https://github.com/bestK/grafana-webhook-converter).

![#](https://raw.githubusercontent.com/bestK/grafana-webhook-converter/master/doc/custom.png)

**Comment by taladar** [comment]
The fact that this has not been done in 8 years despite overwhelming numbers of users asking for it with flimsy excuses about complexity of producing valid JSON in templates is honestly the kind of thing that makes me rethink if Grafana even has a place in our technology stack at all (currently new to it and evaluating). Just throw out your JSON body you have hard-coded somewhere, make a field "Body MIME Type", another "Body Template" and submit that as the HTTP Body of the webhook. That is maybe an hour of work. By all means make another ticket for your advanced use cases that prevent valid JSON and leave that open for another 10 years but this is ridiculous.

**Comment by SeaweedbrainCY** [comment]
@taladar this is a totally inappropriate comment on an open source and free to host project.

If you believe you can implement this feature in one hour, feel free to open a PR.

Otherwise, like all of us that are very interested in this feature, seek or develop a workaround. 
Some workaround are even in this feed. 
I personally created a small fast-api docker container that listen for grafana webhook calls (with the hardcoded json), extract, enrich and call the final webhook. 

It works fine and doesn't require much work or ressources.

**Comment by taladar** [comment]
Running an entire separate workaround application for this issue adds some ridiculous complexity for the fact that their webhook implementation is more complex than it needs to be. It literally has to do less than it does now.

**Comment by grobinson-grafana** [comment]
The main issue with using Go templates to generate JSON for webhooks is it puts the burden on users to write templates that always produce valid JSON. For example, the following template produces invalid JSON:

```
[
	{{ range .Alerts }}
	{
		"name": "{{ index .Labels "alertname" }}"
	},
	{{ end }}
]
```

because there is a trailing comma:

```
[
	
	{
		"name": "foo"
	},
	
]
```

You instead need to write:

```
[
	{{ range $i, $v := .Alerts }}
	{{ if ne $i 0 }},{{end }}{
		"name": "{{ index .Labels "alertname" }}"
	}
	{{ end }}
]
```

If we instead use something like jsonnet then we need to be careful because jsonnet is a turing complete language which means you can write non-terminating programs. We don't want to allow users to write templates with infinite loops and never receive notifications.

These are just two examples of the nuances here.

I promise if we could build this feature in an hour we would have done so. But once we ship a feature we need to support it, and just throwing Go templates at the problem is going to create an ease of use problem for a lot of users.

In the mean time, Grafana webhooks are a superset of Prometheus Alertmanager, one of the most popular open source projects there is in this space. It is and always will be much better to integrate with Alertmanager then the other way around.

**Comment by taladar** [comment]
I really don't see how the usability for your users is any better if they need to build custom adapter applications instead of working around the comma issue. I also feel "Webhooks" is the wrong label for that option if what you really mean is "Prometheus Alertmanager Webhooks", basically you are incompatible with any other application that does the same, i.e. also claims to support Webhooks but requires a fixed format for the body.

**Comment by grobinson-grafana** [comment]
> I really don't see how the usability for your users is any better if they need to build custom adapter applications instead of working around the comma issue.

I agree that it's a pain – but I can promise you the pain of writing correct JSON with a template is even worse. I an unfortunate enough to have direct experience of writing both YAML and JSON in Jinja when I used to be an SRE using saltstack. This is not something I want to push onto users because it's an awful experience. Even if it doesn't seem like it, writing a Lambda function, Cloudflare Worker or equivalent is much easier.

> basically you are incompatible with any other application that does the same, i.e. also claims to support Webhooks but requires a fixed format for the body.

That's quite normal though? A webhook is just a callback that we send when something happens (an alert is fired or resolved). I believe if you were to receive webhooks from Pagerduty, for example, it would be the same. Your software have to understand the [Pagerduty webhook format](https://developer.pagerduty.com/docs/db0fa8c8984fc-overview) rather than the other way around.

**Comment by grobinson-grafana** [comment]
> I really don't see how the usability for your users is any better if they need to build custom adapter applications instead of working around the comma issue.

I agree that it's a pain – but I can promise you the pain of writing correct JSON with a template is even worse. I am unfortunate enough to have direct experience of writing both YAML and JSON in Jinja when I used to be an SRE using saltstack. This is not something I want to push onto users because it's an awful experience. Even if it doesn't seem like it, writing a Lambda function, Cloudflare Worker or equivalent is much easier.

> basically you are incompatible with any other application that does the same, i.e. also claims to support Webhooks but requires a fixed format for the body.

That's quite normal though? A webhook is just a callback that we send when something happens (an alert is fired or resolved). I believe if you were to receive webhooks from Pagerduty, for example, it would be the same. Your software has to understand the [Pagerduty webhook format](https://developer.pagerduty.com/docs/db0fa8c8984fc-overview) rather than the other way around.

**Comment by LwtsdeFrisch** [comment]
+1

**Comment by tremendoustj** [comment]
💯 
This is a must have thing.
